ENH: Make IEC transform logic independent from MRML

markgerencser opened the pull request:
Make it independent in order to be able to use it separately. Remove all the MRML based functions and move it to vtkSlicerBeamsModuleLogic, vtkSlicerRoomsEyeViewModuleLogic.

markgerencser committed:
ENH: Make IEC transform logic independent from MRML

Make IEC transform logic independent in order to be able to use it separately.
- Remove all the MRML based functions and move them to vtkSlicerBeamsModuleLogic, vtkSlicerRoomsEyeViewModuleLogic.
- Functions were moved to follow the REV -> Beams -> IEC dependency.
- All beams related functions from IEC (UpdateBeamTransform and UpdateIECTransformsFromBeam) were moved to BeamsModuleLogic. Reason: used in Beams and MRML.

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -182,9 +182,8 @@ void vtkSlicerBeamsModuleLogic::UpdateTransformForBeam(vtkMRMLRTBeamNode* beamNo
   }
 
   //TODO: Use one IEC logic in a private scene for all beam transform updates?
```
Remove this please.

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -209,9 +208,7 @@ void vtkSlicerBeamsModuleLogic::UpdateTransformForBeam(vtkMRMLScene* beamSequenc
   }
 
   //TODO: Use one IEC logic in a private scene for all beam transform updates?
```
Remove

cpinter replied:
Actually please add a todo instead of this:
`//TODO: beamSequenceScene is not used! Check what it is used for. The error message suggests that it is the main MRML scene, but its name says otherwise.

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -299,3 +296,195 @@ void vtkSlicerBeamsModuleLogic::ProcessMRMLNodesEvents(vtkObject* caller, unsign
+  fixedReferenceToRASTransform->Concatenate(vtkTransform::SafeDownCast(patientSupportRotationToFixedReferenceTransformNode->GetTransformFromParent()));
+
+  fixedReferenceToRasTransformNode->SetAndObserveTransformToParent(fixedReferenceToRASTransform);
+}
```
Add empty line at the end of the file (always)

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.h:
```diff
@@ -80,6 +105,9 @@ class VTK_SLICER_BEAMS_LOGIC_EXPORT vtkSlicerBeamsModuleLogic :
   vtkSlicerMLCPositionLogic* MLCPositionLogic;
+
+private:
+  vtkSmartPointer<vtkSlicerIECTransformLogic> IecLogic = vtkSmartPointer<vtkSlicerIECTransformLogic>::New();
```
Don't use `vtkSmartPointer`, use `vtkNew` (you even have it included although it seems not to be used). See examples in Slicer source code (search for .h files containing "vtkNew")

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -299,3 +296,195 @@ void vtkSlicerBeamsModuleLogic::ProcessMRMLNodesEvents(vtkObject* caller, unsign
+
+  // Set transform to beam node
+  beamTransformNode->SetAndObserveTransformToParent(beamLinearTransform);
+}
```
Empty line needed

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -299,3 +296,195 @@ void vtkSlicerBeamsModuleLogic::ProcessMRMLNodesEvents(vtkObject* caller, unsign
+
+  return vtkMRMLLinearTransformNode::SafeDownCast(
+    this->GetMRMLScene()->GetFirstNodeByName(this->IecLogic->GetTransformNameBetween(fromFrame, toFrame).c_str()));
+}
```
Empty line needed. Please check everywhere. And in general, always make sure that the coding style matches the rest of the code. New code should not be incompliant. Thanks

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.cxx:
```diff
@@ -299,3 +296,195 @@ void vtkSlicerBeamsModuleLogic::ProcessMRMLNodesEvents(vtkObject* caller, unsign
+  }
+
+  // Make sure the transform hierarchy is set up
+  this->IecLogic->BuildIECTransformHierarchy();
```
Are you sure this is needed at every single update?

cpinter commented on Beams/Logic/vtkSlicerBeamsModuleLogic.h:
```diff
@@ -60,6 +64,27 @@ class VTK_SLICER_BEAMS_LOGIC_EXPORT vtkSlicerBeamsModuleLogic :
 
+public:
+  /// Update parent transform node of a given beam from the IEC transform hierarchy and the beam parameters
+  void UpdateBeamTransform(vtkMRMLRTBeamNode* beamNode);
```
Explain in the commit message which function was moved from where to where and why.

cpinter commented on Beams/Logic/vtkSlicerIECTransformLogic.cxx:
```diff
@@ -97,471 +93,103 @@ void vtkSlicerIECTransformLogic::PrintSelf(ostream& os, vtkIndent indent)
-  }
-}
+  return this->GetCoordinateSystemsMap()[fromFrame] + "To" + this->GetCoordinateSystemsMap()[toFrame] + "Transform";
+}
```
Empty line missing

cpinter commented on Beams/Logic/vtkSlicerIECTransformLogic.h:
```diff
@@ -118,60 +116,55 @@ class VTK_SLICER_BEAMS_LOGIC_EXPORT vtkSlicerIECTransformLogic : public vtkMRMLA
   void UpdatePatientSupportRotationToFixedReferenceTransform(double patientSupportRotationAngleDeg);
 
-protected:
+public:
```
You have many `public:`s in quick succession. Please organize a bit better

cpinter commented on DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx:
```diff
@@ -1655,10 +1655,9 @@ vtkMRMLLinearTransformNode* vtkSlicerDrrImageComputationLogic::UpdateImageTransf
 
   // Update transforms in IEC logic from beam node parameters
-  iecLogic->UpdateIECTransformsFromBeam(beamNode);
+  BeamsLogic->UpdateIECTransformsFromBeam(beamNode);
```
What is this `BeamsLogic` variable? Is it local? Then start with small letter. Is it member? Then need `this->` before. Same with `REVLogic` below

cpinter commented on DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.h:
```diff
@@ -198,6 +199,8 @@ class VTK_SLICER_DRRIMAGECOMPUTATION_MODULE_LOGIC_EXPORT vtkSlicerDrrImageComput
   /// Beams logic instance
   vtkSlicerBeamsModuleLogic* BeamsLogic;
+  /// Rooms eye view logic instance
+  vtkSlicerRoomsEyeViewModuleLogic* REVLogic;
```
I think these can be accessed dynamically in every function using the new feature in the slicer application logic class.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -389,6 +389,28 @@ vtkSlicerRoomsEyeViewModuleLogic::~vtkSlicerRoomsEyeViewModuleLogic()
+  // Transforms
+  os << indent << "Transforms:" << std::endl;
+  vtkSmartPointer<vtkMatrix4x4> matrix = vtkSmartPointer<vtkMatrix4x4>::New();
+  for (auto& transformPair : IECLogic->vtkSlicerIECTransformLogic::GetIecTransforms())
```
Please do not have variables like `IECLogic` just like this in the code. It is an error that can mean two different things and make the code less readable. See similar comment above.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -423,13 +443,162 @@ void vtkSlicerRoomsEyeViewModuleLogic::BuildRoomsEyeViewTransformHierarchy()
     vtkErrorMacro("BuildRoomsEyeViewTransformHierarchy: Invalid MRML scene");
     return;
   }
+  vtkSlicerBeamsModuleLogic* beamsLogic = vtkSlicerBeamsModuleLogic::SafeDownCast(this->GetModuleLogic("Beams"));
```
Where I said you can access logic classes dynamically I meant this. No need to define them as members I think.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -423,13 +443,162 @@ void vtkSlicerRoomsEyeViewModuleLogic::BuildRoomsEyeViewTransformHierarchy()
+  }
+
+  // Organize transform nodes into hierarchy based on IEC Standard 61217 (duplicate IEC logic in MRML).
+  // Set and observe individual transforms from IEC logic to the MRML nodes.
```
Explain it here that although we have the concatenated transforms in the IEC logic, we want to define a MRML transform hierarchy because that is how we can set them to the model nodes that are the RT machine parts.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -494,16 +661,16 @@ vtkSlicerRoomsEyeViewModuleLogic::LoadTreatmentMachine(vtkMRMLRoomsEyeViewNode*
     {
-    vtkErrorMacro("LoadTreatmentMachine: Failed to load treatment machine descriptor file '" << descriptorFilePath << "'");
-    return std::vector<TreatmentMachinePartType>();
+      vtkErrorMacro("LoadTreatmentMachine: Failed to load treatment machine descriptor file '" << descriptorFilePath << "'");
```
It's not here where you need to add spaces. It is the `{` from where you need to delete two spaces.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -1003,8 +1183,16 @@ void vtkSlicerRoomsEyeViewModuleLogic::UpdateGantryToFixedReferenceTransform(vtk
 
   this->IECLogic->UpdateGantryToFixedReferenceTransform(parameterNode->GetGantryRotationAngle());
+  vtkMRMLLinearTransformNode* gantryToFixedReferenceTransformNode = beamsLogic->GetTransformNodeBetween(vtkSlicerIECTransformLogic::Gantry, vtkSlicerIECTransformLogic::FixedReference);
+  gantryToFixedReferenceTransformNode->Modified();
```
Explain in a comment (inline, at the end of the line) why you need to call `Modified`. Please do it for every `Modified` call in this commit (if there are several one after the other it's enough to do this for the first one but needed in different places and functions)

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -1645,3 +1866,167 @@ std::string vtkSlicerRoomsEyeViewModuleLogic::GetStateForPartType(std::string pa
 }
+
+//-----------------------------------------------------------------------------
+bool vtkSlicerRoomsEyeViewModuleLogic::GetTransformNodeBetween(vtkSlicerIECTransformLogic::CoordinateSystemIdentifier fromFrame, vtkSlicerIECTransformLogic::CoordinateSystemIdentifier toFrame,
```
Is it possible that you have this function more than once? I see a `beamsLogic->GetTransformNodeBetween` call a bit above, and now the same function is defined here. Keep only the one in the lower level class if they are duplicated.

cpinter commented on RoomsEyeView/Logic/vtkSlicerRoomsEyeViewModuleLogic.cxx:
```diff
@@ -1645,3 +1866,167 @@ std::string vtkSlicerRoomsEyeViewModuleLogic::GetStateForPartType(std::string pa
+  {
+    return false;
+  }
+}
```
End of file empty line needed

cpinter committed:
ENH: Refactor IEC logic to serve Beams functions

- Beams module is able to compose its transform without the IEC transform MRML nodes, only by using the vtkTransform's in the IEC logic
- Allow setting IEC logic to Beams for testing
- Rename test in Beams module from IEC to Beams (the full IEC test will be performed by the REV module)

cpinter committed:
BUG: Fix UpdateRASRelatedTransforms in Beams logic to use the IEC logic instead of the MRML tree

cpinter committed:
ENH: Move GetTransformNodeBetween from Beams to REV; Fix RAS related transform update

markgerencser committed:
ENH: Add get beams logic function and setter for beamslogic

Automated tests in REV needed a setter for beams logic. Get module logic function is used frequently in REV logic for beams, got replaced by function which returns correct beams logic from application or externally set beams logic when running tests.

markgerencser committed:
WIP

cpinter committed:
BUG: Fix beam couch angle computation

Since in the current version what is stationary in RAS is the table top instead of the treatment machine body, the couch angle does not need to be applied directly wrt to the room (FixedReference), but wrt to the table top. Hence the couch angle of the beams is inverted.

cpinter commented:
Integrated as 270e1d19d9eedd09e2822182963cce7200e988a8
